### PR TITLE
Networking dashboard to display per VPC and per VPC peering group limits

### PR DESCRIPTION
**Pull request by aurelienlegrand**
Hi team, 

Proposing to add this tool to monitor GCP networking limits (such Internal forwarding Rules per VPC and per VCP peering group). This is not possible today in Google Console or via APIs and this tool should be very helpful for our customers.

Feedback is welcome as there are probably a lot of improvements to be done on that code but I need help :)

Aurélien

### CODE REVIEW

File: examples/cloud-operations/network-dashboard/cloud-function/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+regex
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/cloud-function/requirements.txt:1]: I'd pin the all the packages listed here to specific versions

File: examples/cloud-operations/network-dashboard/main.tf [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+################################################
+
+# Create an app engine application (required for Cloud Scheduler)
+resource "google_app_engine_application" "scheduler_app" {
+  project = module.project-monitoring.project_id
+  # "europe-west1" is called "europe-west" and "us-central1" is "us-central" for App Engine, see https://cloud.google.com/appengine/docs/locations
+  location_id = var.region == "europe-west1" || var.region == "us-central1" ? substr(var.region, 0, length(var.region) - 1) : var.region
+}
+
+# Create a storage bucket for the Cloud Function's code
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/main.tf:90]: nit: the `cloud-function` module does the same thing you're doing here with GCS, the zip file, etc. No need to change but it would make the code a bit simpler

File: examples/cloud-operations/network-dashboard/main.tf [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+  http_target {
+    http_method = "POST"
+    uri         = google_cloudfunctions_function.function_quotas.https_trigger_url
+    # We could pass useful data in the body later
+    body = base64encode("{\"foo\":\"bar\"}")
+  }
+}
+
+# TODO: How to secure Cloud Function invokation? Not   member = "allUsers" but specific Scheduler service Account?
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/main.tf:164]: I think this is an important point. What exactly do you want to do with the SA?

File: examples/cloud-operations/network-dashboard/main.tf [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  limit_l7_ppg_list        = tolist(var.limit_l7_ppg)
+  limit_l7_ppg             = join(",", local.limit_l7_ppg_list)
+}
+
+################################################
+# Monitoring project creation                  #
+################################################
+
+module "project-monitoring" {
+  source          = "github.com/GoogleCloudPlatform/cloud-foundation-fabric///modules/project"
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/main.tf:44]: I'd ping all these modules to specific versions too.

Comment (juliocc) [reply examples/cloud-operations/network-dashboard/main.tf:44]: And I think you have an extra slash extra slash before "modules/". This one in particular should be ""github.com/GoogleCloudPlatform/cloud-foundation-fabric//modules/project""

**Review by juliocc — CHANGES_REQUESTED** [review]
Very nice. I left a few comments

File: examples/cloud-operations/network-dashboard/main.tf (hunk 4 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf]: Done in my 2nd commit, thanks.

File: examples/cloud-operations/network-dashboard/cloud-function/requirements.txt (hunk 1 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/cloud-function/requirements.txt]: Done in my 2nd commit, thanks.

File: examples/cloud-operations/network-dashboard/main.tf (hunk 2 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf:90]: I would have to test that, do you handle the europe-west1 region being called "europe-west" by AppEngine? I know I had to include an ugly test for 2 corner cases.

File: examples/cloud-operations/network-dashboard/main.tf (hunk 3 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf:164]: For now this is an HTTP trigger but it might make sense to have a pub/sub trigger instead and refactor that part. The goal is to ensure Cloud Scheduler can trigger the Cloud Function but also users are able to trigger it if required (for testing purposes).

Comment (ludoo) [reply examples/cloud-operations/network-dashboard/main.tf:164]: If you look at the quota monitor example we have, there's a pubsub trigger. I would avoid using HTTP at all costs.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+from google.cloud import monitoring_v3
+from googleapiclient import discovery
+from google.api import metric_pb2 as ga_metric
+import time
+import os
+import google.api_core
+import re
+import random
+
+#
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:26]: can you remove this empty comment?

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
+from googleapiclient import discovery
+from google.api import metric_pb2 as ga_metric
+import time
+import os
+import google.api_core
+import re
+import random
+
+# 
+monitored_projects_list = os.environ.get("monitored_projects_list").split(",")  # list of projects from which function will get quotas information
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:27]: These are constants, they should be all uppercase. Can you also move inline comments ton their own line, and make sure you observe line length limits? In Fabric we standardize on yapf for formatting, you should use it as it will trigger PR check errors if you don't.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 7]
```diff
@@ -0,0 +1,791 @@
...
+def quotas(request):
+  '''
+    Cloud Function Entry point, called by the scheduler.
+
+      Parameters:
+        request: for now, the Cloud Function is triggered by an HTTP trigger and this request correspond to the HTTP triggering request.
+      Returns:
+        'Function executed successfully'
+  '''
+  global client, interval
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:51]: Globals are an antipattern in Python. What you probably want is to have a `get_client` function, and each individual function to accept an optional `client` arg.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+  client, interval = create_client()
+
+  instance_metric = create_gce_instances_metrics()
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:54]: All the different create metrics function should probably be a single function, reading external metric definitions from a yaml or similar file. Or at least a constant map describing the different metrics. What you are doing is creating a ton of boilerplate code for no real reason.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+
+from google.cloud import monitoring_v3
+from googleapiclient import discovery
+from google.api import metric_pb2 as ga_metric
+import time
+import os
+import google.api_core
+import re
+import random
+
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:25]: Your function does not seem to have a clear entry point. Please use a `main()` function, and the standard `if __name__ == `__main__` invocation. Use the existing cloud functions in `examples/cloud_operations` as a guide on how Python should look in Fabric, please.

**Review by ludoo — CHANGES_REQUESTED** [review]
There's quite a bit of changes needing in the Python. As I noted in one of the comments, you need proper linting, and also refer to existing Python Cloud Functins in this repo for structure.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py (hunk 8 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/cloud-function/main.py]: This is fixed in my last commit, thanks.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py (hunk 6 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/cloud-function/main.py]: Done, thanks.

File: examples/cloud-operations/network-dashboard/main.tf (hunk 2 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf]: I now use the cloud-function module.

File: examples/cloud-operations/network-dashboard/main.tf (hunk 3 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf:158]: I changed it to a pub sub trigger.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 10]
```diff
@@ -0,0 +1,622 @@
+#
```
Comment (ludoo) [inline examples/cloud-operations/network-dashboard/cloud-function/main.py:1]: Can you format this file with the same command used here

https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/34f8d05f8dc807467860701967e4e9d1f3d1313b/.github/workflows/linting.yml#L74

We have a check that verifies Python scripts for compliant formatting, and it will fail on this PR unless you format this file.

**Review by ludoo — CHANGES_REQUESTED** [review]

File: examples/cloud-operations/network-dashboard/cloud-function/main.py (hunk 10 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/cloud-function/main.py:1]: I ran yapf and applied the fixes.

File: examples/cloud-operations/network-dashboard/cloud-function/main.py [hunk 11]
```diff
@@ -0,0 +1,622 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from google.cloud import monitoring_v3
+from googleapiclient import discovery
+from google.api import metric_pb2 as ga_metric
+import time
```
[the remark this answers is not in the document]
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/cloud-function/main.py:20]: Done manually, it wasn't done by yapf (maybe I missed some option?)

File: examples/cloud-operations/network-dashboard/main.tf [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    event    = "google.pubsub.topic.publish"
+    resource = module.pubsub.topic.id
+    retry    = null
+  }
+}
+
+# TODO: How to secure Cloud Function invokation? Not   member = "allUsers" but specific Scheduler service Account?
+# Maybe "service-YOUR_PROJECT_NUMBER@gcp-sa-cloudscheduler.iam.gserviceaccount.com"?
+
+resource "google_cloudfunctions_function_iam_member" "invoker" {
```
[the remark this answers is not in the document]
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/main.tf:161]: Correct, thanks.

**Review by ludoo — APPROVED** [review]

**Review by ludoo — APPROVED** [review]

File: examples/cloud-operations/network-dashboard/tests/test.tf [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+  allow_stopping_for_update = true
+}
+
+resource "google_compute_instance" "test-vm-hub1" {
+  project      = module.project-svc-hub.project_id
+  name         = "test-vm-hub1"
+  machine_type = "f1-micro"
+  zone         = var.zone
+
+  tags = ["${var.region}"]
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/tests/test.tf:253]: tags = [var.region]

File: examples/cloud-operations/network-dashboard/main.tf [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+  time_zone = "UTC"
+
+  pubsub_target {
+    topic_name = module.pubsub.topic.id
+    data       = base64encode("test")
+  }
+}
+
+# Random ID to re-deploy the Cloud Function  with every Terraform run
+resource "random_pet" "random" {
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/main.tf:108]: I might be missing something but I don't think you need this. The CF module will zip all the source files and redploy if the md5 of the resulting file changes. Do you really need a random bucket id every run?

File: examples/cloud-operations/network-dashboard/tests/test.tf [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+resource "google_folder" "test-net-dash" {
+  display_name = "test-net-dash"
+  parent       = "organizations/${var.organization_id}"
+}
+
+##### Creating host projects, VPCs, service projects #####
+
+module "project-hub" {
+  source          = "github.com/GoogleCloudPlatform/cloud-foundation-fabric//modules/project?ref=v14.0.0"
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/tests/test.tf:25]: I know I said before you should ping the module version but I thought we were in a different repo. You should really use relative paths here (i.e. "../../../../modules/project"). Sorry...

File: examples/cloud-operations/network-dashboard/tests/variables.tf [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+  description = "The organization id for the associated services"
+}
+
+variable "billing_account" {
+  description = "The ID of the billing account to associate this project with"
+}
+
+variable "prefix" {
+  description = "Customer name to use as prefix for resources' naming"
+  default     = "net-dash"
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/tests/variables.tf:27]: we don't usually provide defaults for the prefix

File: examples/cloud-operations/network-dashboard/variables.tf [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+  description = "The ID of the billing account to associate this project with"
+}
+
+variable "prefix" {
+  description = "Customer name to use as prefix for resources' naming"
+  default     = "net-dash"
+}
+
+# Not used for now as I am creating the monitoring project in my main.tf file
+variable "monitoring_project_id" {
```
Comment (juliocc) [inline examples/cloud-operations/network-dashboard/variables.tf:31]: If not used, then remove (or comment) this variable to avoid confusion

**Review by juliocc — CHANGES_REQUESTED** [review]

File: examples/cloud-operations/network-dashboard/tests/variables.tf (hunk 16 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/tests/variables.tf:27]: good point, thanks!

File: examples/cloud-operations/network-dashboard/variables.tf (hunk 17 above)
Comment (aurelienlegrand) [reply examples/cloud-operations/network-dashboard/variables.tf:31]: Done.

**Review by juliocc — APPROVED** [review]
